### PR TITLE
Move `Gate.label` to parent `Instruction` class

### PR DESCRIPTION
**Pull request by chriseclectic**
### Summary

Moves `_label` attribute, and `label` getter/setter from `Gate` to its parent class `Instruction`.

### Details and comments

There isn't really any reason labels should be specific to `Gate` subclasses rather than part for the base `Instruction` class. This allows setting custom labels for general instructions using the `inst.label` setter.

### CODE REVIEW

**Review by kdk — COMMENTED** [review]
Can you also include the `circuit.name` as the instruction label in `circuit.to_instruction()`?

**Comment by chriseclectic** [comment]
I think the remaining failing tests require a change in the Aer snapshot instruction

File: qiskit/circuit/instruction.py [hunk 1]
```diff
@@ -64,6 +64,7 @@ def __init__(self, name, num_qubits, num_clbits, params, duration=None, unit="dt
                 list of parameters
             duration (int or float): instruction's duration. it must be integer if ``unit`` is 'dt'
             unit (str): time unit of duration
+            label: An optional label for identifying the instruction.
```
Comment (Cryoris) [inline qiskit/circuit/instruction.py:67]: Could you add a type hint here?

File: qiskit/circuit/instruction.py [hunk 2]
```diff
@@ -275,13 +281,36 @@ def assemble(self):
...
+        """Set instruction label to name
+
+        Args:
+            name (str or None): label to assign instruction
+
+        Raises:
+            TypeError: name is not string or None.
+        """
+        if isinstance(name, (str, type(None))):
+            self._label = name
```
Comment (Cryoris) [inline qiskit/circuit/instruction.py:310]: Not necessarily something to change, but I'm a bit confused why pylint doesn't complain about `self._label` not being defined in init. Maybe that's due to the workaround until Aer's next release, which may define `self._label`.

**Review by Cryoris — COMMENTED** [review]
Two minor comments otherwise LGTM 🙂

File: qiskit/circuit/instruction.py (hunk 1 above)
Comment (Cryoris) [reply qiskit/circuit/instruction.py:67]:
```suggestion
            label (str or None): An optional label for identifying the instruction.
```

File: qiskit/circuit/instruction.py (hunk 2 above)
Comment (kdk) [reply qiskit/circuit/instruction.py:310]: My best guess is that because `self._label` might be set (on L87), pylint doesn't consider this an issue.

**Review by kdk — APPROVED** [review]

**Pull request merged**